### PR TITLE
Module/sigprofiler/1.0

### PR DESCRIPTION
**Pull request by ppararaj**
# Pull Request Checklists

**Important:** When opening a pull request, keep only the applicable checklist and delete all other sections.

### Required

- [x] I used the cookiecutter template and updated the placeholder rules.

- [x] The snakemake rules follow the [design guidelines](https://lcr-modules.readthedocs.io/en/latest/for_developers.html#module-rules). 

  - [x] All references to the `rules` object (_e.g._ for input files) are wrapped with `str()`.

- [x] Every rule in the module is either listed under `localrules` or has the `threads` and `resources` directives.

- [x] Input and output files are being symlinked into the `CFG["inputs"]` and `CFG["outputs"]` subdirectories, respectively.

- [x] I updated the final target rule (`*_all`) to include every output rule.

- [x] I explained important module design decisions in `CHANGELOG.md`.

- [x] I tested the module on real data for all supported `seq_type` values.

- [x] I updated the `default.yaml` configuration file to provide default values for each rule in the module snakefile.

- [x] I did not set any global wildcard constraints. Any/all wildcard constraints are set on a per-rule basis. 

- [x] I ensured that all symbolic links are relative and self-contained (_i.e._ do not point outside of the repository).

- [x] I replaced every value that should (or might need to) be updated in the default configuration file with `__UPDATE__`.

- [x] I recursively searched for all comments containing `TODO` to ensure none were left. For example:

  ```bash
  grep -r TODO modules/<module_name>/1.0
  ```

### If applicable

- [x] I added more granular output subdirectories.

- [ ] I added rules to the `reference_files` workflow to generate any new reference files.

- [ ] I added subdirectories with large intermediate files to the list of `scratch_subdirectories` in the `default.yaml` configuration file.

- [x] I updated the list of available wildcards for the input files in the `default.yaml` configuration file.

## Checklist for Updated Module

Important! If you are updating the module version, ensure the previous version of the module is restored from master.
If you want to restore a deleted file or directory from the remote master, you can use `git checkout origin/master path/to/file`,
then a `git commit` will ensure that file is tracked on your branch again.
Example:
```
mv modules/strelka/1.1 modules/strelka/1.2
git checkout origin/master modules/strelka/1.1
```

### CODE REVIEW

File: modules/sigprofiler/1.0/sigprofiler.smk [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+rule _sigprofiler_all:
+    input:
+        expand(
+            expand(
+                CFG["dirs"]["outputs"] + "cosmic_sigs/{seq_type}--{genome_build}/{file}.COSMIC_{{type}}.De_Novo_map_to_COSMIC_{{type}}.csv",
+            zip,
+            seq_type = CFG["mafs"]["seq_type"],
+            genome_build = CFG["mafs"]["projected_build"],
+            file = CFG["mafs"]["filename"]),
+         type = ['SBS96', 'ID83', 'DBS78'])
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/sigprofiler.smk]: type should be set in the config or something, don't you think?

Comment (ppararaj) [reply modules/sigprofiler/1.0/sigprofiler.smk]: I've specified type as a config parameter. I've also added the ability to specify other available k-mers.

File: modules/sigprofiler/1.0/src/python/run_generator.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import argparse
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/src/python/run_generator.py:1]: Please document what this script actually does with a few lines of in-line comments.

File: modules/sigprofiler/1.0/src/python/run_extractor.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import argparse
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/src/python/run_extractor.py:1]: Document what this script does with a few lines of in-line comments

File: modules/sigprofiler/1.0/sigprofiler.smk [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+# Symlinks the input files into the module results directory (under '00-inputs/')
+rule _sigprofiler_input_maf:
+    input:
+        maf = CFG["inputs"]["maf"]
+    output:
+        maf = CFG["dirs"]["inputs"] + "matrices/{seq_type}--{genome_build}/{sample_set}/{sample_set}.maf"
+    run:
+        op.relative_symlink(input.maf, output.maf)
+
+
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/sigprofiler.smk]: Explain what this rule is actually doing with a comment at the top

File: modules/sigprofiler/1.0/sigprofiler.smk [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    "DBS2976" : 20
+}
+
+##### RULES #####
+
+rule _install_sigprofiler_genome:
+    output:
+        complete = CFG["dirs"]["inputs"] + "sigprofiler_prereqs/{genome_build}.installed"
+    params:
+        ref = lambda w: {"grch37":"GRCh37", "hg19":"GRCh37",
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/sigprofiler.smk]: This needs to handle hs37d5 and eventually other genomes. Try to think of a way to generalize this.

File: modules/sigprofiler/1.0/sigprofiler.smk [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        $(awk 'BEGIN {{OFS=FS=","}} $1 ~ "*" {{S=substr($1,1,length($1)-1); print S+2}}' {input.stat})
+        {params.nmf_repl} {params.norm} {params.nmf_init} {threads}
+        """)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _sigprofiler_output_tsv:
+    input:
+        decomp = str(rules._sigprofiler_run_extract.output.decomp)
+    output:
+        decomp = CFG["dirs"]["outputs"] + "cosmic_sigs/{seq_type}--{genome_build}/{sample_set}.COSMIC_{type}.De_Novo_map_to_COSMIC_{type}.csv"
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/sigprofiler.smk]: Is genome_build here still actually genome_build or is it the projection?

File: modules/sigprofiler/1.0/sigprofiler.smk [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        $(awk 'BEGIN {{OFS=FS=","}} $1 ~ "*" {{S=substr($1,1,length($1)-1); if (S-2<1) {{print 1}} else {{print S-2}}}}' {input.stat})
+        $(awk 'BEGIN {{OFS=FS=","}} $1 ~ "*" {{S=substr($1,1,length($1)-1); print S+2}}' {input.stat})
+        {params.nmf_repl} {params.norm} {params.nmf_init} {threads}
+        """)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _sigprofiler_output_tsv:
+    input:
+        decomp = str(rules._sigprofiler_run_extract.output.decomp)
+    output:
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/sigprofiler.smk:222]: Can you output a .contents file too, like the examples I showed you?

File: workflows/reference_files/2.4/reference_files.smk [hunk 8]
```diff
@@ -381,3 +381,37 @@ rule get_mutect2_small_exac:
...
+    shell:
+        "pip install SigProfilerMatrixGenerator && touch {output.complete}"
+
+rule install_sigprofiler_genome:
+    input:
+        str(rules.install_sigprofiler_matrix_generator.output.complete)
+    output:
+        complete = "downloads/sigprofiler_prereqs/{genome_build}.installed"
+    params:
+        ref = lambda w: {"grch37":"GRCh37", "hg19":"GRCh37",
```
Comment (rdmorin) [inline workflows/reference_files/2.4/reference_files.smk]: Same comment regarding the additional genomes not handled here (there are likely going to be more)

**Review by rdmorin — CHANGES_REQUESTED** [review]

File: modules/sigprofiler/1.0/sigprofiler.smk (hunk 6 above)
Comment (ppararaj) [reply modules/sigprofiler/1.0/sigprofiler.smk]: This refers to the projection build. See the expand call in rule `_sigprofiler_all`

File: modules/sigprofiler/1.0/src/python/run_generator.py (hunk 2 above)
Comment (ppararaj) [reply modules/sigprofiler/1.0/src/python/run_generator.py:1]: Added comments

File: modules/sigprofiler/1.0/src/python/run_extractor.py (hunk 3 above)
Comment (ppararaj) [reply modules/sigprofiler/1.0/src/python/run_extractor.py:1]: Added comments and argparse help

File: modules/sigprofiler/1.0/sigprofiler.smk (hunk 4 above)
Comment (ppararaj) [reply modules/sigprofiler/1.0/sigprofiler.smk]: Added rule comments

**Comment by ppararaj** [comment]
- [x] Rule comments
- [x] Script comments
- [x] Additional genome support
- [x] contents file generation
- [x] parameterize sigprofiler options

File: modules/sigprofiler/1.0/src/python/run_extractor.py [hunk 9]
```diff
@@ -24,17 +25,17 @@ def main():
...
-    parser.add_argument('ref')
-    parser.add_argument('context_type', default = '96,DINUC,ID')
-    parser.add_argument('exome', default = False, type = bool)
-    parser.add_argument('minimum_signatures', default = 1, type = int)
-    parser.add_argument('maximum_signatures', default = 25, type = int)
-    parser.add_argument('nmf_replicates', default = 100 , type = int)
-    parser.add_argument('matrix_normalization', default = 'gmm')
-    parser.add_argument('nmf_init', default = 'nndsvd_min')
-    parser.add_argument('cpu', default = -1, type = int)
+    parser.add_argument('input_data', help = 'Path to matrix that will undergo NMF.')
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/src/python/run_extractor.py:44]: Have you set it up so these options are available to be tweaked by the user via the config?

File: modules/sigprofiler/1.0/src/python/run_extractor.py [hunk 10]
```diff
@@ -2,9 +2,10 @@
 from SigProfilerExtractor import sigpro as sig
 import os
 
-os.environ["OPENBLAS_NUM_THREADS"] = "1"
+os.environ["OPENBLAS_NUM_THREADS"] = "1" # ulimit fix
```
[the remark this answers is not in the document]
Comment (ppararaj) [reply modules/sigprofiler/1.0/src/python/run_extractor.py:5]: Doesn't seem to affect the multithreading of NMF. However not setting this option causes errors on all machines (from what I remember).

File: modules/sigprofiler/1.0/config/default.yaml [hunk 11]
```diff
@@ -30,10 +70,10 @@ lcr-modules:
 
         pairing_config:
             genome:
-                run_paired_tumours: True
-                run_unpaired_tumours_with: "unmatched_normal"
+                run_paired_tumours: False
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/config/default.yaml]: Why was this changed? Won't this make it harder to find the input files?

File: modules/sigprofiler/1.0/sigprofiler.smk [hunk 12]
```diff
@@ -89,22 +88,17 @@ max_sigs = {
     "DBS2976" : 20
 }
 
-##### RULES #####
```
Comment (rdmorin) [inline modules/sigprofiler/1.0/sigprofiler.smk]: We really shouldn't be hard-coding references into a module like this. Can't this be set in the config?

File: modules/sigprofiler/1.0/config/default.yaml (hunk 11 above)
Comment (ppararaj) [reply modules/sigprofiler/1.0/config/default.yaml]: I was trying to figure out a way to get around an error but I can just include general.yaml config file. The error had to do with an unmatched normal not being provided.

**Review by lkhilton — APPROVED** [review]

**Review by lkhilton — APPROVED** [review]
